core/templates/dev/head/components/common-layout-directives/navigation-bars/top-navigation-bar.directive.ts

```diff
@@ -287,14 +204,102 @@ angular.module('oppia').directive('topNavigationBar', [
             DebouncerService.debounce(truncateNavbar, 500);
 
-          // The function needs to be run after i18n. A timeout of 0 appears to
-          // run after i18n in Chrome, but not other browsers. The function will
-          // check if i18n is complete and set a new timeout if it is not. Since
-          // a timeout of 0 works for at least one browser, it is used here.
+          // The function needs to be run after i18n. A timeout of 0 appears
+          // to run after i18n in Chrome, but not other browsers. The function
+          // will check if i18n is complete and set a new timeout if it is
+          // not. Since a timeout of 0 works for at least one browser,
+          // it is used here.
           $timeout(truncateNavbar, 0);
           $scope.$on('searchBarLoaded', function() {
             $timeout(truncateNavbar, 100);
           });
+          ctrl.$onInit = function() {
+            ctrl.isModerator = null;
+            ctrl.isAdmin = null;
+            ctrl.isTopicManager = null;
+            ctrl.isSuperAdmin = null;
+            ctrl.userIsLoggedIn = null;
+            ctrl.username = '';
+            UserService.getUserInfoAsync().then(function(userInfo) {
+              if (userInfo.getPreferredSiteLanguageCode()) {
+                $translate.use(userInfo.getPreferredSiteLanguageCode());
+              }
+              ctrl.isModerator = userInfo.isModerator();
+              ctrl.isAdmin = userInfo.isAdmin();
+              ctrl.isTopicManager = userInfo.isTopicManager();
+              ctrl.isSuperAdmin = userInfo.isSuperAdmin();
+              ctrl.userIsLoggedIn = userInfo.isLoggedIn();
+              ctrl.username = userInfo.getUsername();
+              if (ctrl.username) {
+                ctrl.profilePageUrl = UrlInterpolationService.interpolateUrl(
+                  '/profile/<username>', {
+                    username: ctrl.username
+                  });
+              }
+
+              if (ctrl.userIsLoggedIn) {
+                // Show the number of unseen notifications in the navbar and
+                // page title, unless the user is already on the dashboard page.
+                $http.get('/notificationshandler').then(function(response) {
+                  var data = response.data;
+                  if ($window.location.pathname !== '/') {
+                    ctrl.numUnseenNotifications = data.num_unseen_notifications;
+                    if (ctrl.numUnseenNotifications > 0) {
+                      $window.document.title = (
+                        '(' + ctrl.numUnseenNotifications + ') ' +
+                        $window.document.title);
+                    }
+                  }
+                });
+              }
+            });
+            UserService.getProfileImageDataUrlAsync().then(function(dataUrl) {
+              ctrl.profilePictureDataUrl = dataUrl;
+            });
+            ctrl.currentUrl = window.location.pathname.split('/')[1];
```

Comment:
Perhaps put all these variable declarations and "simple" initializations at the top, and have all the various async calls come below that. That would be neater, I think? Currently it all reads a bit messily.

Reply:
Done.